File: src/lib/videoDiscovery.ts
import { getPool, parseVideoEvent, type ParsedVideo, ALL_VIDEO_KINDS } from "./nostr";
import type { Filter, Event } from "nostr-tools";

// Aggregator relays that index the broadest content
const DISCOVERY_RELAYS = [
  "wss://relay.damus.io",
  "wss://nos.lol",
  "wss://relay.primal.net",
  "wss://relay.snort.social",
  "wss://nostr.wine",
  "wss://offchain.pub",
  "wss://nostr21.com",
];

// Cache engagement scores in memory and localStorage
const ENGAGEMENT_CACHE_KEY = "videorelay_engagement";
const ENGAGEMENT_CACHE_TTL = 1000 * 60 * 15; // 15 min

interface EngagementScore {
  likes: number;
  dislikes: number;
  comments: number;
  reposts: number;
  zapAmount: number;
  zapCount: number;
  score: number; // composite
  ratioed: boolean;
}

let engagementCache: Map<string, EngagementScore> = new Map();
let cacheTimestamp = 0;

function loadEngagementCache() {
  try {
    const raw = localStorage.getItem(ENGAGEMENT_CACHE_KEY);
    if (!raw) return;
    const { data, timestamp } = JSON.parse(raw);
    if (Date.now() - timestamp > ENGAGEMENT_CACHE_TTL) return;
    engagementCache = new Map(Object.entries(data));
    cacheTimestamp = timestamp;
  } catch {}
}

function saveEngagementCache() {
  try {
    const obj: Record<string, EngagementScore> = {};
    engagementCache.forEach((v, k) => { obj[k] = v; });
    localStorage.setItem(ENGAGEMENT_CACHE_KEY, JSON.stringify({
      data: obj,
      timestamp: Date.now(),
    }));
  } catch {}
}

/**
 * Compute engagement scores for a set of video events.
 * Fetches reactions (kind 7), reposts (kind 6), replies (kind 1), and zaps (kind 9735).
 */
async function fetchEngagementScores(videoIds: string[]): Promise<Map<string, EngagementScore>> {
  const pool = getPool();
  const scores = new Map<string, EngagementScore>();

  const zero = (): EngagementScore => ({
    likes: 0, dislikes: 0, comments: 0, reposts: 0,
    zapAmount: 0, zapCount: 0, score: 0, ratioed: false,
  });

  // Initialize all with zero scores
  for (const id of videoIds) {
    if (engagementCache.has(id)) {
      scores.set(id, engagementCache.get(id)!);
    } else {
      scores.set(id, zero());
    }
  }

  // Only fetch for uncached IDs
  const uncached = videoIds.filter(id => !engagementCache.has(id));
  if (uncached.length === 0) return scores;

  // Batch in chunks
  const chunkSize = 50;
  for (let i = 0; i < uncached.length; i += chunkSize) {
    const chunk = uncached.slice(i, i + chunkSize);

    try {
      // Fetch reactions (kind 7), reposts (kind 6, 16), comments (kind 1111, 1), and zaps (kind 9735)
      // Use broader relay set for engagement queries — zaps especially need good coverage
      const engagementRelays = [...new Set([
        ...DISCOVERY_RELAYS,
        "wss://relay.nostr.band",
        "wss://relay.damus.io",
        "wss://nostr.wine",
        "wss://cache2.primal.net/v1",
      ])].slice(0, 6);

      const [reactions, comments, zaps] = await Promise.allSettled([
        pool.querySync(engagementRelays, {
          kinds: [7, 6, 16],
          "#e": chunk,
          limit: 2000,
        }),
        pool.querySync(engagementRelays, {
          kinds: [1111, 1],
          "#e": chunk,
          limit: 1000,
        }),
        pool.querySync(engagementRelays, {
          kinds: [9735],
          "#e": chunk,
          limit: 2000,
        }),
      ]);

      // Count reactions (likes/dislikes) and reposts
      if (reactions.status === "fulfilled") {
        for (const event of reactions.value) {
          const eTag = event.tags.find(t => t[0] === "e");
          if (!eTag) continue;
          const videoId = eTag[1];
          const entry = scores.get(videoId);
          if (!entry) continue;

          if (event.kind === 7) {
            // NIP-25: "+" or "" = like, "-" = dislike
            if (event.content === "-") {
              entry.dislikes++;
            } else {
              entry.likes++;
            }
          } else if (event.kind === 6 || event.kind === 16) {
            entry.reposts++;
          }
        }
      }

      // Count comments
      if (comments.status === "fulfilled") {
        for (const event of comments.value) {
          const eTag = event.tags.find(t => t[0] === "e");
          if (!eTag) continue;
          const videoId = eTag[1];
          const entry = scores.get(videoId);
          if (entry) entry.comments++;
        }
      }

      // Count zaps and amounts
      if (zaps.status === "fulfilled") {
        for (const event of zaps.value) {
          const eTag = event.tags.find(t => t[0] === "e");
          if (!eTag) continue;
          const videoId = eTag[1];
          const entry = scores.get(videoId);
          if (!entry) continue;

          entry.zapCount++;
          const descTag = event.tags.find(t => t[0] === "description");
          if (descTag) {
            try {
              const desc = JSON.parse(descTag[1]);
              const amountTag = desc.tags?.find((t: string[]) => t[0] === "amount");
              if (amountTag) {
                entry.zapAmount += Math.floor(parseInt(amountTag[1]) / 1000);
              }
            } catch {}
          }
        }
      }
    } catch (err) {
      console.warn("Engagement fetch error:", err);
    }
  }

  // Compute composite scores and cache
  for (const [id, entry] of scores) {
    // Ratioed: dislikes > likes * 2 → exclude from trending
    entry.ratioed = entry.dislikes > entry.likes * 2 && entry.dislikes > 3;

    // Scoring: likes*0.5 - dislikes*2 + (zapAmount/1000)*5 + comments*0.3 + reposts*1
    entry.score = (entry.likes * 0.5)
      - (entry.dislikes * 2)
      + (entry.zapAmount / 1000) * 5
      + (entry.comments * 0.3)
      + (entry.reposts * 1);

    engagementCache.set(id, entry);
  }
  saveEngagementCache();

  return scores;
}

/**
 * Deep crawl for video content across all time periods.
 * Returns videos with real engagement scores.
 */
export async function discoverPopularVideos(
  relays: string[],
  options: {
    timePeriod?: "today" | "week" | "month" | "year" | "all";
    limit?: number;
    sortBy?: "trending" | "zaps";
  } = {}
): Promise<ParsedVideo[]> {
  const { timePeriod = "all", limit = 300, sortBy = "trending" } = options;
  loadEngagementCache();

  const pool = getPool();
  const allEvents: Event[] = [];
  const seen = new Set<string>();

  const addEvents = (events: Event[]) => {
    for (const e of events) {
      if (!seen.has(e.id)) {
        seen.add(e.id);
        allEvents.push(e);
      }
    }
  };

  const now = Math.floor(Date.now() / 1000);
  const day = 86400;

  // Use discovery relays + user's relays for maximum coverage
  const queryRelays = [...new Set([...DISCOVERY_RELAYS, ...relays])].slice(0, 8);

  // Build time windows based on selected period
  let timeWindows: { since?: number; until?: number }[];

  if (timePeriod === "all") {
    // Deep crawl across all time — multiple windows for broader coverage
    timeWindows = [
      { since: now - 7 * day },                           // last week
      { since: now - 30 * day, until: now - 7 * day },    // 1-4 weeks ago
      { since: now - 90 * day, until: now - 30 * day },   // 1-3 months ago
      { since: now - 180 * day, until: now - 90 * day },  // 3-6 months ago
      { since: now - 365 * day, until: now - 180 * day }, // 6-12 months ago
      { until: now - 365 * day },                          // older than 1 year
    ];
  } else {
    const sincePeriod = (() => {
      switch (timePeriod) {
        case "today": return now - day;
        case "week": return now - 7 * day;
        case "month": return now - 30 * day;
        case "year": return now - 365 * day;
      }
    })();
    timeWindows = [{ since: sincePeriod }];
  }

  // Query each time window across relays in parallel
  const promises = timeWindows.flatMap(window =>
    queryRelays.map(async relay => {
      try {
        const filter: Filter = {
          kinds: ALL_VIDEO_KINDS,
          limit: Math.ceil(limit / timeWindows.length),
          ...window,
        };
        const events = await Promise.race([
          pool.querySync([relay], filter),
          new Promise<Event[]>((_, reject) => setTimeout(() => reject(new Error("timeout")), 10000)),
        ]);
        addEvents(events);
      } catch {}
    })
  );

  await Promise.allSettled(promises);

  // Parse all events into videos
  const parsed = allEvents
    .map(parseVideoEvent)
    .filter((v): v is ParsedVideo => v !== null);

  const deduped = dedupeVideos(parsed);

  if (deduped.length === 0) return [];

  // Fetch real engagement scores
  const scores = await fetchEngagementScores(deduped.map(v => v.id));

  // Filter ratioed content and apply scores
  const nowSecs = Date.now() / 1000;
  const filtered = deduped.filter(v => {
    const engagement = scores.get(v.id);
    if (engagement?.ratioed) return false;
    if (engagement) {
      v.zapCount = engagement.zapAmount;
    }
    return true;
  });

  if (sortBy === "zaps") {
    // Most Zapped: pure zap amount ranking
    filtered.sort((a, b) => {
      const zapA = scores.get(a.id)?.zapAmount || 0;
      const zapB = scores.get(b.id)?.zapAmount || 0;
      if (zapA !== zapB) return zapB - zapA;
      // Tiebreak by zap count
      const countA = scores.get(a.id)?.zapCount || 0;
      const countB = scores.get(b.id)?.zapCount || 0;
      return countB - countA;
    });
    // Filter out videos with zero zaps for Most Zapped
    return filtered.filter(v => {
      const e = scores.get(v.id);
      return e && (e.zapAmount > 0 || e.zapCount > 0);
    });
  }

  // Trending: composite score + time decay
  filtered.sort((a, b) => {
    const scoreA = scores.get(a.id)?.score || 0;
    const scoreB = scores.get(b.id)?.score || 0;

    // Time decay: boost recent content (168 hours = 1 week)
    const ageHoursA = (nowSecs - a.publishedAt) / 3600;
    const ageHoursB = (nowSecs - b.publishedAt) / 3600;
    const decayA = Math.max(0, 168 - ageHoursA) / 168;
    const decayB = Math.max(0, 168 - ageHoursB) / 168;

    const finalA = scoreA + decayA * 10;
    const finalB = scoreB + decayB * 10;
    return finalB - finalA;
  });

  return filtered;
}

function dedupeVideos(videos: ParsedVideo[]): ParsedVideo[] {
  const seen = new Set<string>();
  return videos.filter((v) => {
    if (seen.has(v.id)) return false;
    seen.add(v.id);
    return true;
  });
}

File: android/app/src/main/java/com/videorelay/app/data/nostr/VideoDiscovery.kt
package com.videorelay.app.data.nostr

import android.content.Context
import android.util.Log
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import com.videorelay.app.domain.model.Video
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.first
import kotlinx.serialization.json.*
import javax.inject.Inject
import javax.inject.Singleton

private val Context.discoveryStore by preferencesDataStore(name = "discovery_cache")

data class EngagementScore(
    val likes: Int = 0,
    val dislikes: Int = 0,
    val comments: Int = 0,
    val reposts: Int = 0,
    val zapAmount: Long = 0,  // in sats
    val zapCount: Int = 0,
    val score: Double = 0.0,
    val ratioed: Boolean = false,
)

enum class SortMode { Trending, MostZapped }

@Singleton
class VideoDiscovery @Inject constructor(
    @ApplicationContext private val context: Context,
    private val relayPool: RelayPool,
) {
    companion object {
        private const val TAG = "VideoDiscovery"
        private val KEY_CACHE = stringPreferencesKey("engagement_cache")
        private const val CACHE_TTL_MS = 15 * 60 * 1000L // 15 min

        // Aggregator relays with broadest content index
        val DISCOVERY_RELAYS = listOf(
            "wss://relay.nostr.band",
            "wss://relay.primal.net",
            "wss://relay.damus.io",
            "wss://nos.lol",
            "wss://relay.snort.social",
            "wss://nostr.wine",
            "wss://offchain.pub",
        )
    }

    private val engagementCache = mutableMapOf<String, EngagementScore>()
    private var cacheTimestamp = 0L

    /**
     * Fetch engagement scores (likes, dislikes, comments, reposts, zaps)
     * for a list of video IDs. Mirrors web app's fetchEngagementScores().
     */
    suspend fun fetchEngagementScores(videoIds: List<String>): Map<String, EngagementScore> {
        if (videoIds.isEmpty()) return emptyMap()

        // Load cache
        if (engagementCache.isEmpty()) loadCache()

        val now = System.currentTimeMillis()
        val cacheValid = (now - cacheTimestamp) < CACHE_TTL_MS

        val scores = mutableMapOf<String, EngagementScore>()
        val uncached = mutableListOf<String>()

        for (id in videoIds) {
            val cached = if (cacheValid) engagementCache[id] else null
            if (cached != null) {
                scores[id] = cached
            } else {
                scores[id] = EngagementScore()
                uncached.add(id)
            }
        }

        if (uncached.isEmpty()) return scores

        // Query engagement relays in chunks of 50
        val engagementRelays = DISCOVERY_RELAYS.take(6)

        uncached.chunked(50).forEach { chunk ->
            try {
                coroutineScope {
                    val reactionsJob = async {
                        relayPool.query(
                            engagementRelays,
                            NostrFilter(kinds = listOf(7, 6, 16), tags = mapOf("#e" to chunk), limit = 2000),
                            timeoutMs = 5000,
                        )
                    }
                    val commentsJob = async {
                        relayPool.query(
                            engagementRelays,
                            NostrFilter(kinds = listOf(1111, 1), tags = mapOf("#e" to chunk), limit = 1000),
                            timeoutMs = 5000,
                        )
                    }
                    val zapsJob = async {
                        relayPool.query(
                            engagementRelays,
                            NostrFilter(kinds = listOf(NostrConstants.ZAP_RECEIPT_KIND), tags = mapOf("#e" to chunk), limit = 2000),
                            timeoutMs = 5000,
                        )
                    }

                    val reactions = reactionsJob.await()
                    val comments = commentsJob.await()
                    val zaps = zapsJob.await()

                    // Count reactions
                    for (event in reactions) {
                        val videoId = event.getTag("e") ?: continue
                        val entry = scores[videoId] ?: continue
                        scores[videoId] = when (event.kind) {
                            7 -> if (event.content == "-") entry.copy(dislikes = entry.dislikes + 1)
                                 else entry.copy(likes = entry.likes + 1)
                            6, 16 -> entry.copy(reposts = entry.reposts + 1)
                            else -> entry
                        }
                    }

                    // Count comments
                    for (event in comments) {
                        val videoId = event.getTag("e") ?: continue
                        val entry = scores[videoId] ?: continue
                        scores[videoId] = entry.copy(comments = entry.comments + 1)
                    }

                    // Count zaps
                    for (event in zaps) {
                        val videoId = event.getTag("e") ?: continue
                        val entry = scores[videoId] ?: continue
                        var zapAmount = 0L
                        val descTag = event.getTag("description")
                        if (descTag != null) {
                            try {
                                val desc = Json.parseToJsonElement(descTag).jsonObject
                                val tags = desc["tags"]?.jsonArray
                                val amountTag = tags?.firstOrNull {
                                    it.jsonArray.firstOrNull()?.jsonPrimitive?.content == "amount"
                                }
                                amountTag?.let {
                                    zapAmount = it.jsonArray[1].jsonPrimitive.content.toLong() / 1000
                                }
                            } catch (_: Exception) {}
                        }
                        scores[videoId] = entry.copy(
                            zapCount = entry.zapCount + 1,
                            zapAmount = entry.zapAmount + zapAmount,
                        )
                    }
                }
            } catch (e: Exception) {
                Log.w(TAG, "Engagement fetch error: ${e.message}")
            }
        }

        // Compute composite scores
        for ((id, entry) in scores) {
            val ratioed = entry.dislikes > entry.likes * 2 && entry.dislikes > 3
            val score = (entry.likes * 0.5)
                .minus(entry.dislikes * 2.0)
                .plus((entry.zapAmount / 1000.0) * 5.0)
                .plus(entry.comments * 0.3)
                .plus(entry.reposts * 1.0)
            scores[id] = entry.copy(ratioed = ratioed, score = score)
            engagementCache[id] = scores[id]!!
        }

        cacheTimestamp = System.currentTimeMillis()
        saveCache()
        return scores
    }

    /**
     * Sort videos by trending or most zapped using full engagement scores.
     * Mirrors web app's discoverPopularVideos().
     */
    suspend fun sortByEngagement(
        videos: List<Video>,
        mode: SortMode,
        timePeriodSeconds: Long = Long.MAX_VALUE,
    ): List<Video> {
        if (videos.isEmpty()) return emptyList()

        val now = System.currentTimeMillis() / 1000

        // Apply time filter
        val filtered = if (timePeriodSeconds == Long.MAX_VALUE) videos
        else videos.filter { it.publishedAt >= now - timePeriodSeconds }

        if (filtered.isEmpty()) return emptyList()

        val scores = fetchEngagementScores(filtered.map { it.id })

        // Filter ratioed content
        val clean = filtered.filter { scores[it.id]?.ratioed != true }

        return when (mode) {
            SortMode.MostZapped -> {
                clean
                    .filter { (scores[it.id]?.zapAmount ?: 0) > 0 || (scores[it.id]?.zapCount ?: 0) > 0 }
                    .sortedByDescending { v ->
                        val e = scores[v.id]
                        (e?.zapAmount ?: 0L) * 1000 + (e?.zapCount ?: 0)
                    }
                    .ifEmpty { clean.sortedByDescending { v -> scores[v.id]?.zapCount ?: 0 } }
            }
            SortMode.Trending -> {
                clean.sortedByDescending { v ->
                    val e = scores[v.id]
                    val score = e?.score ?: 0.0
                    // Time decay: boost content from last week
                    val ageHours = ((now - v.publishedAt) / 3600.0).coerceAtLeast(1.0)
                    val decay = (168.0 - ageHours.coerceAtMost(168.0)) / 168.0
                    score + decay * 10.0
                }
            }
        }
    }

    private suspend fun loadCache() {
        try {
            val prefs = context.discoveryStore.data.first()
            val raw = prefs[KEY_CACHE] ?: return
            val json = Json.parseToJsonElement(raw).jsonObject
            val ts = json["ts"]?.jsonPrimitive?.longOrNull ?: 0L
            val data = json["data"]?.jsonObject ?: return
            cacheTimestamp = ts
            for ((id, obj) in data) {
                val e = obj.jsonObject
                engagementCache[id] = EngagementScore(
                    likes = e["l"]?.jsonPrimitive?.intOrNull ?: 0,
                    dislikes = e["d"]?.jsonPrimitive?.intOrNull ?: 0,
                    comments = e["c"]?.jsonPrimitive?.intOrNull ?: 0,
                    reposts = e["r"]?.jsonPrimitive?.intOrNull ?: 0,
                    zapAmount = e["za"]?.jsonPrimitive?.longOrNull ?: 0L,
                    zapCount = e["zc"]?.jsonPrimitive?.intOrNull ?: 0,
                    score = e["s"]?.jsonPrimitive?.doubleOrNull ?: 0.0,
                    ratioed = e["rt"]?.jsonPrimitive?.booleanOrNull ?: false,
                )
            }
        } catch (_: Exception) {}
    }

    private suspend fun saveCache() {
        try {
            val data = buildJsonObject {
                put("ts", cacheTimestamp)
                put("data", buildJsonObject {
                    engagementCache.entries.take(500).forEach { (id, e) ->
                        put(id, buildJsonObject {
                            put("l", e.likes); put("d", e.dislikes)
                            put("c", e.comments); put("r", e.reposts)
                            put("za", e.zapAmount); put("zc", e.zapCount)
                            put("s", e.score); put("rt", e.ratioed)
                        })
                    }
                })
            }
            context.discoveryStore.edit { it[KEY_CACHE] = data.toString() }
        } catch (_: Exception) {}
    }
}
